jfr/src/main/java/io/netty/jfr/JfrChannelHandler2.java

```diff
@@ -0,0 +1,539 @@
+/*
+ * Copyright 2019 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.jfr;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandler.Sharable;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import jdk.jfr.Category;
+import jdk.jfr.Enabled;
+import jdk.jfr.Event;
+import jdk.jfr.Label;
+import jdk.jfr.Name;
+import jdk.jfr.Threshold;
+
+import java.net.SocketAddress;
+
+/**
+ * A {@link ChannelHandler} that creates JDK Flight Recorder events for all Netty channel events/operations.
+ */
+@Sharable
+public final class JfrChannelHandler2 extends ChannelDuplexHandler {
```

Comment:
So why we have two handlers ?

Reply:
I just added this one to show my idea of what the ChannelPromise fix would look like (bad name of the class though). Since it messed up the stacktraces I didn't want to commit it. But maybe I can figure something out to make sure event.commit() is run in the right context